### DISCUSSION TITLE
Filament v2 Updates

### DISCUSSION THREAD

**Discussion by danharrin** [filamentphp/filament#495] [General]
Hi everyone! Hope you're all very well.

After a successful (but still ongoing) consultation, we've got a huge list of features we're planning for Filament v2! Thanks to everyone who has contributed so far. You can check the full list in #408.

I've opened this thread to start open communication with the community about the progress with v2, report challenges we face when building new features, and give you a place to ask questions!

Consider subscribing to this thread to receive any updates.

#### When will this be ready?

Good question! We're currently planning to release v2 in 3 separate stages:

- Form builder [filament/forms](https://github.com/laravel-filament/forms): August / September 2021.
- Table builder [filament/tables](https://github.com/laravel-filament/tables): September / October 2021.
- Admin panel [filament/filament](https://github.com/laravel-filament/filament): November / December 2021.

Many users may not know this, but Filament is actually split into 3 "modules": the form builder, the table builder, and the admin panel. These are technically independent packages from each other already, but usage on their own is undocumented. One of the goals of v2 is to give users the opportunity to use the form and table builders in their own apps without the admin panel. Because of this, and because we want to provide these resources as early as possible, we will be releasing the builders before the admin panel. This will give people a change to test them, iron any bugs out, and familiarise themselves with any breaking changes, to make the transition for admin panel v2 a lot smoother.

Note: when the form and table builders are released, they will not be compatible with admin panel v1. Installing a mixture of v1 and v2 will result in a Composer dependency exception.

#### How easy will it be to upgrade? Will there be many breaking changes?

With v1 of any project, it's inevitable that features could have been created differently in hindsight. This means that some of the changes we will be making will break your existing codebases. Fortunately, along with an upgrade guide, I'm going to look at creating a few automated upgrade tools to make the repetitive parts of the upgrade process easier. And, of course, you're welcome to continue using v1 for as long as it takes you to upgrade your existing apps.

#### How long will v1 be supported for after v2 is released?

This has not been fully decided yet, but I see no reason why I would decide not to release bug fixes for v1 for many months, and security fixes for even longer. The current maintenance burden of Filament v1 is pretty low as its very stable.

**Comment by bernhardh** [comment]
Maybe its a dump question, but why reinvent the wheel and not just use existing stuff. For example for the tables use https://tables.laravel-boilerplate.com/tailwind ? This is Livewire+Tailwind and has almost "all" the features discussed in the other thread?

**Reply by danharrin** [reply]
Yeah this looks like an excellent implementation, however I would much rather maintain our own generator since it will give us complete control over our future roadmap and guarantee to our users that it will be updated. I will look at forking some of the features since it's MIT, thanks for the pointer.

**Comment by danharrin** [comment]
Hey all! I've been working for the past two weeks on v2 of the form builder package, so I thought it would be a nice time to update you all...

I'm over half way done, which I'm feeling great about. The new package includes some exciting new features, including a JSON repeater, block-based page builder, and `spatie/laravel-medialibrary` support, with lots more to come!

I've been updating [the feature checklist](https://github.com/laravel-filament/filament/discussions/408#discussion-3349924) with my progress, if you want to see more details!

As always, drop any questions in the replies if you have them!

Dan

**Comment by danharrin** [comment]
Hello! Time for another update 🎉

I'm getting **very** close to the end of development work for v2 of the forms package. The only tasks left on my todos are:

- [x] Multi-select field.
- [x] Adding filename customisation support to the file upload fields.
- [x] Tabs component.
- [x] Adding relationship support to the select fields.

After that, I need to rebuild the docs website in Laravel (it's currently using Jigsaw), allowing it to support multiple packages. I also need to write the actual docs for the form builder.

I think that I will aim to launch this package and the new docs site during the first half of September, as originally estimated.

Thanks for all the continual support of my work, and if you have any questions drop them below 👇 

Dan

**Comment by danharrin** [comment]
Hey folks!

We're even closer now... I have most of the docs & video tutorials produced now. I think we'll be releasing the `filament/forms` package on Tuesday or Wednesday.

> Please note, using the `filament/forms` package in the same project as `filament/filament` will not be supported until v2 is fully released - in Winter 2021. This is due to namespacing conflicts.

Dan

**Comment by danharrin** [comment]
Hi everyone 👋 

Really pleased with how the form builder release went! Hope you're all enjoying it.

Now to focus on the table builder... this will hopefully will be ready by the end of October.

If you want a reminder of our features for this release check out #495.

Stay tuned!

Dan

**Reply by studiowizjo** [reply]
Hello Dan,
Any further news/estimations when we can expect the new version?

**Reply by danharrin** [reply]
Hey everyone!

Tables v2 has been released (on time? 😉) and I already have a good working version of v2 admin that I'm using at work.

If you'd like a live update on progress with the new features - check #495.

I expect v2 admin to be fully released during the last week of November / first week of December :)

Dan

**Reply by elfeffe** [reply]
Hi @danharrin and is there any way to begin to use it in beta?
I only see the documentation for version 1, and there is some documentation for form and tables but we still need Azmon 2 to test them.
So is there any way to test it all?
Install V2 and then proceed with forms and tables? Will that work?

**Reply by ryangjchandler** [reply]
@elfeffe You can use the tables and forms packages outside of the admin. The admin package is essentially a wrapper around those packages.

**Reply by robertorinaldi-dev** [reply]
@elfeffe 

I am testing version 2.x of admin by copying the updated files (https://github.com/laravel-filament/admin/tree/2.x) to a folder called filament and inserting these lines into composer:
```
"require": {
...
        "filament/filament": "*",
        "filament/forms": "^2.4",
        "filament/tables": "^2.4",
...
    },
...
"repositories": [
        {
            "type": "path",
            "url": "./filament"
        }
    ]
```
Hope it helps!

**Reply by danharrin** [reply]
Eager! 😂

**Reply by danharrin** [reply]
Btw - you can install it with `composer require filament/filament:2.x-dev`.

**Reply by robertorinaldi-dev** [reply]
It was easier! :)
I tried composer require filament/filament:2.x

Thanks!!

**Reply by ericp-mrel** [reply]
I get the following error when trying to install the v2 branch from composer.

![Screenshot from 2021-11-23 10-58-47](https://user-images.githubusercontent.com/63664924/143069595-34710da6-b9a6-4fc7-bc73-d4ac2807afab.png)

**Reply by jwjcmw** [reply]
@ericp-mrel Install the 2.0 Forms and Tables packages separately, they aren't yet included with the Admin package.

**Reply by ericp-mrel** [reply]
@jwjcmw Ah, thanks!

**Comment by tangjh** [comment]
Missing Create button in table list despite overloading both can methods to return true for resource file:

        "filament/filament": "2.x-dev",
        "filament/forms": "^2.4",
        "filament/tables": "^2.4",

    public static function can(string $action, ?Model $record = null): bool
    {
        return true;
    }

    public static function canCreate(): bool
    {
        return true;
    }

    public static function getPages(): array
    {
        return [
            'index' => Pages\ListItemMakes::route('/'),
            'create' => Pages\CreateItemMake::route('/create'),
            'edit' => Pages\EditItemMake::route('/{record}/edit'),
        ];
    }

**Reply by danharrin** [reply]
Just fixed this.

**Reply by MACscr** [reply]
Guessing that change isn't pushed yet? i don't see it the commit log.

**Reply by molcsab** [reply]
It seems that it has to do with user permissions. I've made a Policy for the model and once I have added the following lines, the Create button appeared.

```php
public function create(User $user)
    {
        return $user->role == 'admin';
    }

**Reply by danharrin** [reply]
Sorry guys, I'd pushed it to the dev-main branch instead of 2.x-dev 👍 should be fixed now @tangjh @MACscr @molcsab

**Comment by MACscr** [comment]
That’s odd when no permissions are needed to create or edit, but needed for the button to show?

- Mark

**Reply by danharrin** [reply]
Yup, this was a bug but it's been fixed. I was just missing a ! haha.

**Comment by danharrin** [comment]
Hey guys, just wanted to give you all a quick update as to where I'm at.

I've finished developing every new feature for v2. I've also finished moving all the old documentation over, tweaking it for v2 & writing a simple upgrade guide for v1 users.

I've received some feedback from Discord users who have already tried out the beta, so I need to make a few more changes before we release. These are:

- [x] Improve relation manager performance by only rendering 1 at a time.
- [x] Create another & attach another buttons for resource pages and relation managers (an old v1 feature).
- [x] Add configuration to define the max width of all admin pages.
- [x] Extract more views for the sidebar & top bar to allow them to be customised more easily.
- [x] Add columns to table filter forms, to allow for wider popovers.

I've also got to write documentation for all the new features in v2.

Finally, I've been sent existing v1 admin panels to help me improve the upgrade guide further, in return for a free upgrade PR to v2 :) These will be finished before I release it.

So, I'm aiming to release v2 on **Tuesday, Wednesday or Thursday next week**, depending on how quickly these tasks are finished.

If you're eager to try out v2 right away to provide feedback, you can do so by running `composer require filament/forms filament/tables filament/filament:2.x-dev`. You can find the documentation in its current state [here](https://github.com/laravel-filament/filament/tree/main/packages/admin/docs).

Thanks for all the ongoing support guys!

Dan

**Comment by bezhanSalleh** [comment]
@danharrin the **CreateRecord Hooks** behaves as follow:

![FilamentCreateRecordHooks](https://user-images.githubusercontent.com/10007504/144530027-5a525638-e399-4e47-a42f-8cbd8e421f48.png):

This shouldn't be the intent here but may be it is?

**Reply by danharrin** [reply]
Hey, hooks are not meant to set up redirects, just run logic. To redirect, you'll need to execute a method like `$this->redirect(static::$resource::getUrl('index'));`

**Reply by bezhanSalleh** [reply]
> Hey, hooks are not meant to set up redirects, just run logic. To redirect, you'll need to execute a method like `$this->redirect(static::$resource::getUrl('index'));`

This might sound dumb but bear with me cause I'm kinda confused. Been going through the codebase since you tweeted about it.

If you mean to execute that method inside the hook! 
I tried it and also a variation of it like `return redirect(static::$resource::getUrl('index'));` or redirect to a totally different route. It didn't take.

If not then overriding the `getRedirectUrl()` method does the job perfectly outside the hooks.

I tried every kinda logic i could think of inside hooks and it works smoothly. 

But couldn't get to redirect through a hook if a condition were to fail.

**Reply by danharrin** [reply]
TBH I think it's a Livewire issue somewhere, that does not allow you to redirect from methods not directly called by Livewire. What do you think?

`getRedirectUrl()` works because we're still calling the `redirect()` from the `create()` method.

Btw - what is your use-case for redirecting in hooks?

**Reply by bezhanSalleh** [reply]
hmm... might be a livewire issue, and I did try calling the `redirect()` in all sort of places inside the `create()` method as you can see 🙂 ... even overrod it but it only works as commented out.

~~~
    public function create(): void
    {
        $this->redirect('/posts'); // no redirection - continues
        $this->redirect(route('filament.resources.posts.index')); // no redirection - continues

        $data = $this->form->getState();

        $this->redirect(route('filament.resources.posts.index')); // no redirection - continues

        $resource = static::getResource(); //extra

        $this->redirect(route('filament.resources.posts.index')); // no redirection - continues

        $this->record = static::getModel()::create($data);

        $this->redirect(route('filament.resources.posts.index')); // no redirection - continues

        $this->form->model($this->record)->saveRelationships(); //seems to work after this point

        $this->redirect(route('filament.resources.posts.index'));

        // if ($redirectUrl = $this->getRedirectUrl()) {
        //     $this->redirect($redirectUrl);
        // }

    }
~~~

also tried it in `v1` same thing.

`abort()` does work but it shows the error page based on the code and message; not ideal.
`abort(redirect())` would've worked but then it requires some core changes in `livewire` internals to handle the response.
but if we supress the exception it won't let the user continue but it will remain on the same page with form filled no alerts 🙃.

will dig in some more, after coffee; but for now that's what i got.

**Reply by danharrin** [reply]
Yeah, this must be a Livewire issue then, not Filament :/

**Comment by danharrin** [comment]
And we've released! Thanks for all the support everyone. Enjoy using Filament v2!

**Comment by adibpwd** [comment]
how to retrieve newly created data in function afterCreate()

**Reply by danharrin** [reply]
`$this->record`

**Comment by andrewdwallo** [comment]
One of my fields that I fill out in a form disappears every time I fill in the value. Usually right after I go back to table view. I made sure that the field is both in the form and in the table. And I made sure it is in my database migration table and model with same format and everything. Idk why it is doing that... anything you suggest? Also there is an error or something in my welcome page saying "Property is ignored due to the display. With 'display:block', vertical-align should not be used." css(propertyIgnoredDueToDisplay)... Not sure what that means just needed some advice.. but other than that your plugin is great!
